### ISSUE TITLE
Support dynamic import() in background service worker

### ISSUE THREAD

**Issue by larry-xu** [hober/web-extensions#212]
As documented in https://github.com/w3c/ServiceWorker/issues/1585, dynamically importing modules is not supported in a service worker environment. This limitation applies to background service workers in MV3 as well. Ideally dynamic import() would be supported for service workers generally, but it doesn't look like there has been any progress on that front within the last year. Are there any plans to support this for browser extensions specifically?

This feature would be useful for extensions that require large dependencies that may only be necessary to load in certain contexts. For example, in our extension we dynamically load specific locale files after detecting the user's language.

**Comment by tophf** [comment]
Chromium bug: https://crbug.com/1198822. The only workaround for now is a dynamic `importScripts` but it executes in global scope, not as a module, and it also requires boilerplate duplication of code in SW `install` event.

**Comment by dotproto** [comment]
> Are there any plans to support this for browser extensions specifically?

On the Chrome side we're considering making an extension-specific allowance for dynamic import in module service workers. I believe @wanderview has made public comments to this effect in service worker issues but I don't have links handy at the moment. I don't anticipate that we will be able to explore this further before the end of 2022.

> The only workaround for now is a dynamic importScripts … also requires boilerplate duplication of code in SW install event.

For clarity, service workers cannot call `importScripts()` after it leaves the "install" phase. https://w3c.github.io/ServiceWorker/#importscripts

You should not need to call `importScripts()` from within the `install` event handler. Since this method executes synchronously, you can call it directly from the root scope of a classic service worker or during the service worker's install event. Either works.

**Comment by tophf** [comment]
> For clarity, service workers cannot call importScripts() after it leaves the "install" phase.

Wrong. It can and it works because the essence of the trick is to have all future code present at the install phase. Once all modules were imported in the install event (when the extension was installed) we can import any of these modules dynamically (inside asynchronous code) at any time.

**Comment by hanguokai** [comment]
According to past [meeting minutes](https://github.com/w3c/webextensions/blob/main/_minutes/2022-05-26-wecg.md), I add supportive labels from Chrome and Firefox. I also open an [issue](https://github.com/w3c/ServiceWorker/issues/1700) at Web service worker, just let them know this WECG proposal for extension service worker.

**Comment by jon49** [comment]
This is my current workaround to this issue. Some of the code is probably not needed. But basically, I just use `new Function(...)` to dynamically add code. I use it to have a website in the service worker and stream the HTML from there.

https://github.com/jon49/jon49-sw/blob/0102f5a25f2c0559029e09f0716abb9f57d7c512/lib/routes.middleware.ts#L91
